operator/pkg/psp/emm.go

```diff
@@ -0,0 +1,681 @@
+	GetDCPods() []*corev1.Pod
+	GetNotReadyPodsBootstrappedInDC() []*corev1.Pod
+	GetAllPodsNotReadyInDC() []*corev1.Pod
+	GetPVCsForPod(pod *corev1.Pod) ([]*corev1.PersistentVolumeClaim, error)
+	StartNodeReplace(podName string) error
+	GetInProgressNodeReplacements() []string
+	RemovePod(pod *corev1.Pod) error
+	UpdatePod(pod *corev1.Pod) error
+	IsStopped() bool
+	IsInitialized() bool
+	GetLogger() logr.Logger
+}
+
+type EMMChecks interface {
+	getSelectedNodeNameForPodPVC(podName string) (string, error)
+	getPodNameSetWithVolumeHealthInaccessiblePVC(rackName string) (map[string]bool, error)
+	getRacksWithNotReadyPodsBootstrapped() []string
+	getNodeNameSetForRack(rackName string) map[string]bool
+	getInProgressNodeReplacements() []string
+	IsStopped() bool
+	IsInitialized() bool
+}
+
+type EMMOperations interface {
+	cleanupEMMAnnotations() (bool, error)
+	getNodeNameSetForPlannedDownTime() (map[string]bool, error)
+	getNodeNameSetForEvacuateAllData() (map[string]bool, error)
+	removeAllNotReadyPodsOnEMMNodes() (bool, error)
+	failEMM(nodeName string, failure EMMFailure) (bool, error)
+	performPodReplaceForEvacuateData() (bool, error)
+	removeNextPodFromEvacuateDataNode() (bool, error)
+	removeAllPodsFromPlannedDowntimeNode() (bool, error)
+	startNodeReplace(podName string) error
+}
+
+type EMMService interface {
+	EMMOperations
+	EMMChecks
+	getLogger() logr.Logger
+}
+
+//
+// Util
+//
+func getRackNameSetForPods(pods []*corev1.Pod) map[string]bool {
+	names := map[string]bool{}
+	for _, pod := range pods {
+		names[pod.Labels[api.RackLabel]] = true
+	}
+	return names
+}
+
+func filterPodsByRackName(pods []*corev1.Pod, rackName string) []*corev1.Pod {
+	return utils.FilterPodsWithLabel(pods, api.RackLabel, rackName)
+}
+
+//
+// EMMOperations impl
+//
+type EMMServiceImpl struct {
+	EMMSPI
+}
+
+func (impl *EMMServiceImpl) getSelectedNodeNameForPodPVC(podName string) (string, error) {
+	matchingPods := utils.FilterPodsWithFn(impl.GetDCPods(), func(pod *corev1.Pod) bool {
+		return pod != nil && pod.Name == podName
+	})
+
+	if len(matchingPods) == 0 {
+		return "", nil
+	}
+
+	pod := matchingPods[0]
+
+	pvcs, err := impl.GetPVCsForPod(pod)
+	if err != nil {
+		return "", err
+	}
+
+	for _, pvc := range pvcs {
+		nodeName := utils.GetSelectedNodeNameForPVC(pvc)
+		if nodeName != "" {
+			return nodeName, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (impl *EMMServiceImpl) cleanupEMMAnnotations() (bool, error) {
+	nodes, err := impl.getNodeNameSetForEvacuateAllData()
+	if err != nil {
+		return false, err
+	}
+	nodes2, err := impl.getNodeNameSetForPlannedDownTime()
+	if err != nil {
+		return false, err
+	}
+	nodes = utils.UnionStringSet(nodes, nodes2)
+
+	// Strip EMM failure annotation from pods where node is no longer tainted
+	podsFailedEmm := impl.getPodsWithAnnotationKey(EMMFailureAnnotation)
+	nodesWithPodsFailedEMM := utils.GetNodeNameSetForPods(podsFailedEmm)
+	nodesNoLongerEMM := utils.SubtractStringSet(
```

Comment:
If we ever want to polish things further, we could abstract a stringset library.